gym_ai2thor/envs/ai2thor_env.py

```diff
@@ -66,106 +73,185 @@ def __init__(self, seed=None, config_file='config_files/config_example.json', co
                     self.event = self.controller.step(
                             dict(action=action_str,
-                                 objectId=self.event.metadata['inventoryObjects'][0]['objectId'],
+                                 objectId=object_to_put['objectId'],
                                  receptacleObjectId=interaction_obj['objectId']))
-            elif action_str == 'PickupObject':
+                    self.event.metadata['lastObjectPut'] = object_to_put
+                    self.event.metadata['lastObjectPutReceptacle'] = interaction_obj
+            elif action_str.startswith('Pickup'):
                 closest_pickupable = None
                 for obj in visible_objects:
                     # look for closest object to pick up
                     if obj['pickupable'] and obj['distance'] < distance and \
-                            obj['objectType'] in self.objects['pickupables']:
-                        closest_pickupable = obj
-                if closest_pickupable and not self.event.metadata['inventoryObjects']:
+                            obj['distance'] < self.task.max_object_pickup_crosshair_dist and \
+                            obj['objectType'] in self.allowed_objects['pickupables']:
+                        if self.task.max_object_pickup_euclidean_dist:
+                            euc_distance_to_obj = calculate_euc_distance_between_agent_and_object(
+                                self.event.metadata['agent'], obj)
+                            # print('3D dist: {}. distance to crosshair: {}'.format(
+                            #     euc_distance_to_obj, obj['distance'])) # todo remove
+                            if euc_distance_to_obj < self.task.max_object_pickup_euclidean_dist:
+                                closest_pickupable = obj
+                        else:
+                            closest_pickupable = obj
+                if closest_pickupable: # and not self.event.metadata['inventoryObjects']: # todo option to check for this@?
                     interaction_obj = closest_pickupable
                     self.event = self.controller.step(
                         dict(action=action_str, objectId=interaction_obj['objectId']))
-            elif action_str == 'OpenObject':
+                    self.event.metadata['lastObjectPickedUp'] = interaction_obj
+            elif action_str.startswith('Open'):
                 closest_openable = None
                 for obj in visible_objects:
                     # look for closest closed receptacle to open it
                     if obj['openable'] and obj['distance'] < distance and not obj['isopen'] and \
-                            obj['objectType'] in self.objects['openables']:
+                            obj['objectType'] in self.allowed_objects['openables']:
                         closest_openable = obj
                         distance = closest_openable['distance']
                     if closest_openable:
```

Comment:
This is wrong. If there are two objects to close and the furthest one is first in visible_objects iterable you will close both the furthest and the closest. Moreover, we are executing the close action as many times as remaining objects in visible_objects after the closest_openable. Unindenting this if block will solve the issue. The same applies to "CloseObject" action